lib/msf/core/rhosts_walker.rb

```diff
@@ -0,0 +1,207 @@
+      end
+    end
+
+    def initialize(value = '', datastore = Msf::ModuleDataStore.new(nil))
+      @value = value
+      @datastore = datastore
+    end
+
+    #
+    # Iterate over the valid rhosts datastores. This can be combined Calling `#valid?` beforehand to ensure
+    # that there are no invalid configuration values, as they will be ignored by this method.
+    #
+    # @yield [Msf::DataStore] Yields only _valid_ rhost values.
+    def each(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::DataStore)
+      end
+
+      nil
+    end
+
+    # Count the _valid_ datastore permutations for the current rhosts value. This count will
+    # ignore any invalid values.
+    #
+    # @return [Integer]
+    def count
+      to_enum.count
+    end
+
+    #
+    # Retrieve the list of errors associated with this rhosts walker
+    # @yield [Msf::RhostsWalker::Error] Yields only invalid rhost values.
+    def errors(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::RhostsWalker::Error)
+      end
+
+      nil
+    end
+
+    #
+    # Indicates that the rhosts value is valid and iterable
+    #
+    # @return [Boolean] True if all items are valid, and there are at least some items present to iterate over. False otherwise.
+    def valid?
+      parsed_values = parse(@value, @datastore)
+      parsed_values.all? { |result| result.is_a?(Msf::DataStore) } && parsed_values.count > 0
+    rescue StandardError => e
+      elog('rhosts walker invalid', error: e)
+      false
+    end
+
+    #
+    # Parses the input rhosts string, and yields the possible combinations of datastore values.
+    #
+    # @param value [String] the http string
+    # @param datastore [Msf::Datastore] the datastore
+    # @return [Enumerable<Msf::DataStore|StandardError>] The calculated datastore values that can be iterated over for
+    #   enumerating the given rhosts, or the error that occurred when iterating over the input
+    def parse(value, datastore)
+      Enumerator.new do |results|
+        values = value.to_s.split(', ').map { |line| line.split(' ') }.flatten
+        values.each do |value|
+          if (value =~ %r{^file://(.*)}) || (value =~ /^file:(.*)/)
+            file = Regexp.last_match(1)
+            File.read(file).each_line(chomp: true) do |line|
+              parse(line, datastore).each do |result|
+                results << result
+              end
+            end
+          elsif value.start_with?('smb:')
+            smb_options = parse_smb_uri(value, datastore)
+            Rex::Socket::RangeWalker.new(smb_options['RHOSTS']).each_ip do |ip|
+              results << datastore.merge(
+                smb_options.merge('RHOSTS' => ip, 'TODO_RHOST_SCHEMA_VALUE' => value)
```

Comment:
I liked leaving the original value around for debugging purposes, but I couldn't think of a sane name that describes the intent concisely. Any thoughts? 🤔 

I'm using it later in the error handling too

Reply:
Does this need to be a consistent value for all types? if not maybe use the URI type being processed `SMB_RHOSTS`, `HTTP_RHOSTS`, `CIDR_RHOSTS`

Reply:
I like the proto-specific RHOSTS mechanism, and it goes well with brute related stuff. Would probably want to delineate in `IPV4_HOSTS` and `IPV6_HOSTS` though for actual IP. This opens up things like `TOR_HOSTS` and other fun options down the line (i had to do  "bad things" in order to have bind/rev handlers working properly with tor-space).